validator/client/runner.go

```diff
@@ -17,75 +17,60 @@ import (
 	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
 	"go.opentelemetry.io/otel/trace"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 )
 
-// Time to wait before trying to reconnect with beacon node.
-var backOffPeriod = 10 * time.Second
+const (
+	slotDeadlineThreshold = 2 * time.Second
+)
 
 // Run the main validator routine. This routine exits if the context is
 // canceled.
 //
 // Order of operations:
-// 1 - Initialize validator data
+// 1 - Init validator data
 // 2 - Wait for validator activation
 // 3 - Wait for the next slot start
 // 4 - Update assignments
 // 5 - Determine role at current slot
 // 6 - Perform assigned role, if any
-func run(ctx context.Context, v iface.Validator) {
+func run(ctx context.Context, v iface.Validator) error {
 	cleanup := v.Done
 	defer cleanup()
 
-	headSlot, err := initializeValidatorAndGetHeadSlot(ctx, v)
-	if err != nil {
-		return // Exit if context is canceled.
-	}
-	ss, err := slots.EpochStart(slots.ToEpoch(headSlot + 1))
-	if err != nil {
-		log.WithError(err).Error("Failed to get epoch start")
-		ss = headSlot
-	}
-	startDeadline := v.SlotDeadline(ss + params.BeaconConfig().SlotsPerEpoch - 1)
-	startCtx, startCancel := context.WithDeadline(ctx, startDeadline)
-	if err := v.UpdateDuties(startCtx); err != nil {
-		handleAssignmentError(err, headSlot)
-	}
-	startCancel()
-	eventsChan := make(chan *event.Event, 1)
-	healthTracker := v.HealthTracker()
-	runHealthCheckRoutine(ctx, v, eventsChan)
-
-	accountsChangedChan := make(chan [][fieldparams.BLSPubkeyLength]byte, 1)
-	km, err := v.Keymanager()
-	if err != nil {
-		log.WithError(err).Fatal("Could not get keymanager")
-	}
-	sub := km.SubscribeAccountChanges(accountsChangedChan)
-	// check if proposer settings is still nil
-	// Set properties on the beacon node like the fee recipient for validators that are being used & active.
-	if v.ProposerSettings() == nil {
-		log.Warn("Validator client started without proposer settings such as fee recipient" +
-			" and will continue to use settings provided in the beacon node.")
-	}
-	if err := v.PushProposerSettings(ctx, km, headSlot, true); err != nil {
-		log.WithError(err).Fatal("Failed to update proposer settings")
+	if err := v.Init(ctx); err != nil {
+		if errors.Is(err, context.Canceled) {
+			return nil // Exit if context is canceled.
+		}
+		return errors.Wrap(err, "failed to initialize validator")
 	}
+	tracker := v.HealthTracker()
+	runHealthCheckRoutine(ctx, v)
+	genesisTime := v.GenesisTime()
 	for {
 		select {
 		case <-ctx.Done():
 			log.Info("Context canceled, stopping validator")
-			sub.Unsubscribe()
-			close(accountsChangedChan)
-			return // Exit if context is canceled.
+			return nil // Exit if context is canceled.
 		case slot := <-v.NextSlot():
-			if !healthTracker.IsHealthy(ctx) {
+			if !tracker.IsHealthy(ctx) {
 				continue
 			}
-
 			deadline := v.SlotDeadline(slot)
+			if time.Until(deadline) < slotDeadlineThreshold {
```

Comment:
to avoid that issue with the slot ticker if consumer is slow it will replay old slots.